src/charm.py

```diff
@@ -487,66 +487,94 @@ def _hwcksum_config_matches_pod_config(self) -> bool:
         return existing_content.get("hwcksum") != self._charm_config.enable_hw_checksum
 
+    def _unit_is_non_active_status(self):
+        if not self.unit.is_leader():
+            # NOTE: In cases where leader status is lost before the charm is
+            # finished processing all teardown events, this prevents teardown
+            # event code from running. Luckily, for this charm, none of the
+            # teardown code is necessary to perform if we're removing the
+            # charm.
+            return BlockedStatus("Scaling is not implemented for this charm")
+        try:  # workaround for https://github.com/canonical/operator/issues/736
+            self._charm_config: CharmConfig = CharmConfig.from_charm(charm=self)  # type: ignore[no-redef]  # noqa: E501
+        except CharmConfigInvalidError as exc:
+            return BlockedStatus(exc.msg)
+        if status := self._check_cpu_compatibility():
+            return status
+        if not self._kubernetes_multus.multus_is_available():
+            return BlockedStatus("Multus is not installed or enabled")
+        if not self._bessd_container.can_connect():
+            return WaitingStatus("Waiting for bessd container to be ready")
+        if not self._kubernetes_multus.is_ready():
+            return WaitingStatus("Waiting for Multus to be ready")
+        if not self._bessd_container.exists(path=BESSD_CONTAINER_CONFIG_PATH):
+            return WaitingStatus("Waiting for storage to be attached")
+        if status := self._check_container_availability():
+            return status
+
+    def _check_container_availability(self):
+        if not service_is_running_on_container(self._bessd_container, self._bessd_service_name):
+            return WaitingStatus("Waiting for bessd service to run")
+        if not service_is_running_on_container(self._bessd_container, self._routectl_service_name):
+            return WaitingStatus("Waiting for routectl service to run")
+        if not service_is_running_on_container(
+            self._pfcp_agent_container, self._pfcp_agent_service_name
+        ):
+            return WaitingStatus("Waiting for pfcp agent service to run")
+
+    def _on_collect_unit_status(self, event: CollectStatusEvent):
+        if status := self._unit_is_non_active_status():
+            event.add_status(status)
+        event.add_status(ActiveStatus())
+
     def _on_config_changed(self, event: EventBase):
         """Handler for config changed events."""
-        # workaround for https://github.com/canonical/operator/issues/736
-        try:
+        try:  # workaround for https://github.com/canonical/operator/issues/736
             self._charm_config: CharmConfig = CharmConfig.from_charm(charm=self)  # type: ignore[no-redef]  # noqa: E501
-        except CharmConfigInvalidError as exc:
-            self.model.unit.status = BlockedStatus(exc.msg)
+        except CharmConfigInvalidError:
             return
         if not self.unit.is_leader():
```

Comment:
``` if not self.unit.is_leader():
            return
``` 
Could we do this check in the charm constructor instead of here ?

Reply:
We don't have any guarantee that the unit that was leader when the constructor is instantiated is still leader when config-changed event is triggered. From [SDK](https://juju.is/docs/sdk/leadership):

> Units that hold leader status should not assume they will retain it, a new leader can be elected at any time.